### PR TITLE
add StudyHierarchy for ISIS data, closes #480

### LINKED ISSUE #480: Create new ISIS hierarchy
**Description**:
In the ISIS demo, it was suggested that perhaps a more intuitive/user friendly hierarchy would be `Instrument -> Investigation (via RB Number) -> Dataset -> Datafile`. Additionally, they thought that perhaps it should be togglable like card view is.

This hierarchy should be easier to implement, and should just require configuring the ISIS investigations table to display the correct data and for RB numbers to be more prominent, and for a toggle to switch between which table is linked to from the Instruments table. Trying to keep both functionalities in the Investigations table might be a bit tricky though, or a bit of a mess with the toggle boolean.

**Acceptance criteria**:
- [ ] New ISIS hierarchy implemented
- [ ] Users can switch between old and new ISIS hierarchy

### PR DESCRIPTION
**Pull request by patrick-austin**
## Description
Introduce alternate hierarchy for ISIS data:
- Instrument > Study > Investigation > Dataset > Datafile

This can be accessed by navigating to `/browseStudyHierarchy/...` rather than `/browse/...` which will include facility cycles as before. Navigation within each hierarchy will only link to other pages with the same browse prefix, so if the plugin is registered with the link `/browse(StudyHierarchy)/instrument` then it will only allow the user to navigate to the corresponding pages (though in reality providing a valid web address will work irrespective of which plugin was linked.

This is done with routing, and the addition of a prop to the existing tables/CVs which determines which address to link to next.

Also changed "Visit ID" to "RB Number" for ISIS views and added translations to CV labels.

Note that sorting for the new study table/CV doesn't work for the same reason as mentioned on #460 and ral-facilities/datagateway-api#122. The latter says this should be fixed, but maybe the api url the dev instance points to (http://scigateway-preprod.esc.rl.ac.uk:5000) doesn't have the fix yet?

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Travis build
- [ ] Review changes to test coverage
- [ ] Check navigation between entities works for both hierarchies

## Agile board tracking
connect to #480

### CODE REVIEW

File: packages/datagateway-dataview/public/res/default.json [hunk 1]
```diff
@@ -35,6 +36,13 @@
     "end_date": "End Date",
     "start_date": "Start Date"
   },
+  "studies": {
+    "name": "Name",
+    "rb_number":"RB Number",
```
Comment (louise-davies) [inline packages/datagateway-dataview/public/res/default.json:41]: I'd prefer this to match the ICAT field name - rb_number isn't a field - it's generally the `name` field

File: packages/datagateway-dataview/src/views/card/isis/isisInvestigationsCardView.component.tsx [hunk 2]
```diff
@@ -195,52 +218,52 @@ const ISISInvestigationsCardView = (
...
-        label: 'Description',
+        label: t('investigations.details.summary'),
         dataKey: 'SUMMARY',
         filterComponent: textFilter,
       }}
       information={[
         {
           icon: <Fingerprint />,
-          label: 'Visit ID',
+          label: t('investigations.rb_number'),
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/card/isis/isisInvestigationsCardView.component.tsx:228]: The correct field name for Visit ID is `visit_id` (see old labels) - "RB number" is actually the name field, but again this shouldn't be a translation key name, instead it should be the value of the name key.

File: packages/datagateway-dataview/src/views/card/isis/isisStudiesCardView.component.tsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+      }}
+      description={{
+        label: t('facilitycycles.description'),
+        dataKey: 'STUDY.DESCRIPTION',
+        filterComponent: textFilter,
+      }}
+      information={[
+        {
+          icon: <Fingerprint />,
+          label: t('studies.rb_number'),
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/card/isis/isisStudiesCardView.component.tsx:161]: Again, would prefer the translation key to match the dataKey

File: packages/datagateway-dataview/src/views/card/isis/isisStudiesCardView.component.tsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        content: (studyInvestigation: StudyInvestigation) =>
+          tableLink(
+            `/${pathRoot}/instrument/${instrumentId}/${instrumentChild}/${studyInvestigation.STUDY.ID}/investigation`,
+            studyInvestigation.STUDY.NAME,
+            view
+          ),
+        filterComponent: textFilter,
+      }}
+      description={{
+        label: t('facilitycycles.description'),
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/card/isis/isisStudiesCardView.component.tsx:154]: Some of the labels here all refer to facilitycycles, not studies

File: packages/datagateway-dataview/src/views/card/isis/isisStudiesCardView.component.tsx [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    ),
+  fetchData: (allIds: number[], offsetParams: IndexRange) =>
+    dispatch(
+      fetchStudies({
+        offsetParams,
+        additionalFilters: [
+          {
+            filterType: 'where',
+            filterValue: JSON.stringify({
+              INVESTIGATION_ID: { in: allIds },
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/card/isis/isisStudiesCardView.component.tsx:204]: This is a clever way to add instrument context to a study search! Again, unsure if this is going to be the final way we fetch study info but it's fine for now

File: packages/datagateway-dataview/src/views/table/isis/isisInvestigationsTable.component.tsx [hunk 6]
```diff
@@ -185,7 +211,7 @@ const ISISInvestigationsTable = (
         },
         {
           icon: <FingerprintIcon />,
-          label: t('investigations.visit_id'),
+          label: t('investigations.rb_number'),
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/table/isis/isisInvestigationsTable.component.tsx:214]: Revert this change

File: packages/datagateway-dataview/src/views/table/isis/isisMyDataTable.component.tsx [hunk 7]
```diff
@@ -205,7 +205,7 @@ const ISISMyDataTable = (
         },
         {
           icon: <FingerprintIcon />,
-          label: t('investigations.visit_id'),
+          label: t('investigations.rb_number'),
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/table/isis/isisMyDataTable.component.tsx:208]: Revert

File: packages/datagateway-dataview/src/views/table/isis/isisStudiesTable.component.tsx [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+          cellContentRenderer: (props: TableCellProps) =>
+            tableLink(
+              `/${pathRoot}/instrument/${instrumentId}/${instrumentChild}/${props.rowData.ID}/investigation`,
+              props.rowData.STUDY?.NAME
+            ),
+          filterComponent: textFilter,
+        },
+        {
+          icon: <FingerprintIcon />,
+          label: t('studies.rb_number'),
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/table/isis/isisStudiesTable.component.tsx:129]: Again, the study RB number is the name. Studies don't have visit ids, since one study could contain multiple visits

**Review by louise-davies — CHANGES_REQUESTED** [review]
Looks good overall, mostly just some confusion on what RB numbers map to in ICAT. However, I have noticed that if you click on a Study, what shows up in the breadcrumb isn't the same name as the study I just clicked on. e.g. clicking on `STUDY 375` in the table/card view takes me to `STUDY 7` in the breadcrumbs. Could you investigate what's going on here?

Again, it should be noted that this view requires study objects to exist, so would require creation of studies for every investigation currently in ISIS ICAT. We will need to discuss whether this is the most sensible option later.

Additionally, we need to think about which URL we're hosting this on. (probably worth opening issues to address these two points)

**Comment by patrick-austin** [comment]
The breadcrumb issue was due to using `STUDYINVESTIGATION.ID` for the link in the studies table rather than `STUDYINVESTIGATION.STUDY.ID` which is what we were (correctly) doing in card view. So it wasn't an issue with the breadcrumbs, but the actual navigation. 
I'll revert my mistaken attempts to include the RB Number, and use it as the string for `NAME`/`investigations.isis.name` instead (on ISIS tables/CVs), and include the `PID`/DOI as a field to display.

File: packages/datagateway-dataview/public/res/default.json [hunk 9]
```diff
@@ -72,6 +82,9 @@
         "label": "Publications",
         "reference": "Reference"
       }
+    },
+    "isis": {
+      "name":"RB Number"
```
Comment (louise-davies) [inline packages/datagateway-dataview/public/res/default.json:87]: I'm not exactly a fan of this "isis" section header - the idea would be that the facility would configure `investigation.name` to be either `RB Number` in ISIS, or `Name` for DLS.

File: packages/datagateway-dataview/src/views/table/isis/isisStudiesTable.component.tsx [hunk 10]
```diff
@@ -114,28 +114,28 @@ const ISISStudiesTable = (
...
       columns={[
         {
-          icon: <TitleIcon />,
+          icon: <FingerprintIcon />,
           label: t('studies.name'),
           dataKey: 'STUDY.NAME',
           cellContentRenderer: (props: TableCellProps) =>
             tableLink(
-              `/${pathRoot}/instrument/${instrumentId}/${instrumentChild}/${props.rowData.ID}/investigation`,
+              `/${pathRoot}/instrument/${instrumentId}/${instrumentChild}/${props.rowData.STUDY.ID}/investigation`,
```
Comment (louise-davies) [inline packages/datagateway-dataview/src/views/table/isis/isisStudiesTable.component.tsx:122]: When I load the table and click on the first instrument, when loading the study table I now get an error: `props.rowData.STUDY is undefined` which I presume stems from here. However,  if I load the page directly (e.g. reloading the page once I get the error) it works fine. This error only occurs in Table view, but if I navigate via Card view then switch to Table view then the data loads correctly with no error. It seems as if the data isn't being requested when the route changes?

I think it might be because of the two useEffects required for fetching data in this table.

**Review by louise-davies — CHANGES_REQUESTED** [review]

File: packages/datagateway-dataview/public/res/default.json (hunk 9 above)
Comment (patrick-austin) [reply packages/datagateway-dataview/public/res/default.json:87]: Ah OK, the original issue mentioned making RB numbers more apparent or something like that so I thought I needed to change something to say "RB Number" as part of the change, but it being configurable by the user makes sense. Provided we're displaying the `<entity>.name` for relevant tables/views it should be OK then?

File: packages/datagateway-dataview/src/views/table/isis/isisStudiesTable.component.tsx (hunk 10 above)
Comment (patrick-austin) [reply packages/datagateway-dataview/src/views/table/isis/isisStudiesTable.component.tsx:122]: When navigating directly to studies table `data` is empty, and when swapping from cardview to table we already have `data` populated with `StudyInvestigation`s. In both cases `props.rowData.STUDY` can be called safely.

When navigating from instruments `data` is populated with `Instrument`s which obviously don't have `STUDY` on them, and so it errors (`clearTable` is called straight away but the error still occurs before it can reset `data`).

We don't see the issue with card view as due to the pagination logic, the link isn't rendered straight away, only after we've called `fetchData` once. All other tables use `<entity>.ID` so this isn't an issue normally, as the new table can still access an ID from the old `data` for however long the split second is before `clearTable` hits.

Rather than fundamentally change anything to do with routing or clearing, just changing it to `STUDY?.ID` will ensure we don't error while waiting between the location change and `clearTable`.

**Review by louise-davies — DISMISSED** [review]
The conditional check fixed the problem I was having so this looks good to go 👍

**Comment by patrick-austin** [comment]
Have updated tests to use translated strings, and the routing to use the new ID checks so re-review needed.

**Review by louise-davies — APPROVED** [review]
still looks good 👍
